README.md
# cr-scrollgallery

react component

## Example
<img src="./github/example.jpg" alt="example" />

## Usage

```jsx
cnpm install --save cr-scrollgallery
```

```js
import ScrollGallery from 'cr-scrollgallery'
import 'cr-scrollgallery/assets/index.css'
// 注意在使用的时候不能忘记引入css
// 如果使用cssModule，不需要在此处引入css，而是在对应的less或scss文件引入，查看注意01
  render () {
    const galleryTop = (
      <div style={{border: '1px solid red', height: 300}}></div>
    )
    const arrayElement = () => {
      const one = (
        <div key='one'>
          <div style={{ height: 80}}>tab一</div>
          <div style={{border: '1px solid green', height: 450}}></div>
        </div>
      )
      const two = (
        <div key='two'>
          <div style={{ height: 60}}>tab二</div>
          <div style={{border: '1px solid green', height: 450}}></div>
        </div>
      )
      const three = (
        <div key='three'>
          <div style={{ height: 70}}>tab三</div>
          <div style={{border: '1px solid green', height: 450}}></div>
        </div>
      )

      return [one, two, three]
    }

    return (
      <div>
        <div style={{height: 600, border: '10px solid #21c8be'}}>
          <ScrollGallery
            barTab={['tab一', 'tab二', 'tab三']}
            tabSelect={3}
            galleryTop={galleryTop}
            galleryElements={arrayElement()}
          />
        </div>
      </div>
    )
  }

```

## 注意01
如果使用了cssModule，作为样式的使用方式，在引入css时应特别注意
以And Design Pro使用为例，在对应的页面less中引入css。注意包名前面的波浪线
```less
@import '~cr-scrollgallery/assets/index.css';
```

## 注意02
如果组件传入的是异步元素，因为组件在`componentDidMount`时计算元素的高度导致不准确，因此组件增加了修补措施，在异步完成是调用方法`this.gallery.setInitScrollNum()`
2019-06-14已修复，一般状态下不需要手动触发修补措施
如何使用：
```js
promise.then(() => {
  this.gallery.setInitScrollNum()
})
<ScrollGallery ref={gallery => this.gallery = gallery }></ScrollGallery>
```

## API
name | type | description
-----|------|------------
barTab|array|标题tab
tabSelect|number|默认选中的tab,并且滚动到相应位置
galleryTop|react node|tab bar上部内容
galleryElements|array of react node|tab对应的页面内容
barMinWidth|string/number|设置tabBar最小宽度，合适的值可以使tabBar居中 (例如：`500`,`'500px'`,`'85%'`)
outBoxHeight|number|设置外层 scroll div 高度，默认600

## Development

```
cnpm install
npm start
```

## Example

http://localhost:8020/examples/testOne.html


## License

cr-scrollgallery is released under the MIT license.


HISTORY.md
# History
----

## 1.11.0
将 react-perfect-scrollbar 封装在组件内
scroll div height 需要组件指定
```
<ScrollGallery outBoxHeight={500} />
```

## 1.0.8 / 2019-06-14
1. 修复某些滚动状态无法选中tab标签
2. 修复传入滚动元素顶部元素设置margin导致滚动元素高度计算不准确
3. 修复在某些持续props更新状态下会自动滚动到对应选中tab顶部

## 1.0.5 / 2019-06-12
修改滚动元素更新后元素高度计算问题

## 1.0.4 / 2019-06-09
滚动子元素统一增加外包div元素,并且增加key

## 1.0.2 / 2019-06-06
增加barMinWidth property,传入合适的值可以使tabBar居中

## 1.0.1 / 2019-06-03
修改默认选中，当默认选中第一个tab时，tabBar 不会显现在最顶部

## 1.0.0 / 2019-06-01

cr-scrollgallery is start












src/index.js
import React, { PureComponent, Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import OuiDom from './utils/ouiDomUtils'
import isEqual from 'lodash/isEqual'
import PerfectScrollbar from 'react-perfect-scrollbar'

class ScrollGalleryShow extends PureComponent {

  constructor(props) {
    super(props)
    this.state = {
      tabSelect: props.tabSelect,
      barFix: false,
      scrollBarH: 0,
      barTabH: 0,
      listElementMessage: [],
      scrollTop: 0,
      label: true,
    }
  }

  componentDidMount() {
    const { prefixCls } = this.props
    const barTabRect = this.galleryBox.getBoundingClientRect()
    const elementW = OuiDom.outerWidth(this.gallery)
    const gTopH = OuiDom.outerHeightWithMargin(this.gTop)
    const barTabH = OuiDom.outerHeightWithMargin(this.barTab)
    const scrollBarH = gTopH
    // console.log('顶部高度', gTopH)
    OuiDom.setStyles(this.barTabTwo, {
      ['display']: 'none',
      ['width']: elementW + 'px',
    })
    this.setState({
      scrollBarH,
      barTabH,
    })
    this.setInitScrollNum()
    this.setIndicatorH()
  }

  setIndicatorH = () => {
    // 设置传入元素高度指纹
    const length = this.state.listElementMessage.length
    const indicatorH = this.state.scrollBarH + this.state.listElementMessage[length-1]
    this.setState({
      indicatorH
    })
  }

  setInitScrollNum = () => {
    const { prefixCls } = this.props
    const listElement = this.listBox.querySelectorAll(`.${prefixCls}-child`)
    const listElementArr = Array.from(listElement)
    const elementMessage = []
    // console.log('需要滚动的子元素', listElementArr);
    // 修正第一次检测
    elementMessage.push(0)
    listElementArr.forEach((ele, index, arr) => {
      let height = OuiDom.outerHeightWithMargin(ele)
      let sumH = 0
      for (let i=0; i<=index; i++) {
        sumH += OuiDom.outerHeightWithMargin(arr[i])
        // console.log('元素高度', i, OuiDom.outerHeightWithMargin(arr[i]) )
      }
      elementMessage.push(sumH)
    })
    this.setState({
      listElementMessage: elementMessage,
    }, () => {
      // console.log('子元素高度累加', this.state.listElementMessage)
      if (this.state.tabSelect > 0 && this.state.label) {
        this.setScrollTop(this.state.tabSelect)
      }
    })   
  }
  componentDidUpdate(prevProps, prevState) {
    if (!isEqual(this.props, prevProps)) {
      this.setInitScrollNum()
      this.setIndicatorH()
      if (isEqual(this.state.indicatorH, prevState.indicatorH)) {
        this.setState({
          label: false
        })
      } else {
        this.setState({
          label: true
        }) 
      }   
    }
  }

  handleTabClick = (e, val) => {
    this.setState({
      tabSelect: val
    })
    this.setScrollTop(val)
  }

  setScrollTop = (val) => {
    const gTopH = this.state.scrollBarH
    const elementTop = this.state.listElementMessage[val]
    this.galleryBox.scrollTop = gTopH + elementTop
  }

  handleGalleryScroll = (e) => {
    let scrollTop = e.target.scrollTop
    let gTopH = this.state.scrollBarH
    //向上滚动还是向下滚动
    let scrollUp = true
    scrollTop > this.state.scrollTop ? scrollUp = true : scrollUp = false
    
    if (scrollTop > gTopH ) {
      this.barTab.style.display = 'none'
      this.listBox.style.marginTop = this.state.barTabH + 'px'
      this.barTabTwo.style.display = 'block'
    } else {
      this.listBox.style.marginTop = 0
      this.barTab.style.display = 'block'
      this.barTabTwo.style.display = 'none'
    }

    let listElementMessage = this.state.listElementMessage
    let length = listElementMessage.length

    // 提前多少距离选中tab
    const interval = 50
    const galleryBoxH = OuiDom.outerHeightWithMargin(this.galleryBox)
    
    for (let i=0; i<length; i++) {
      if (scrollUp) {
        // console.log('页面往上滑动')
        let tabSelect
        if (scrollTop < listElementMessage[i]+gTopH-interval) {
          i === 0 ? tabSelect = i : tabSelect = i - 1
          this.setState({
            tabSelect
          })
          break
        }  

      } else {
        // console.log('页面向上滚动的距离', scrollTop)
        // console.log('页面往下滑动', i, listElementMessage[i])
        let tabSelect
        let lastTabSelect
        if (scrollTop < listElementMessage[i]+gTopH-interval) {
          // console.log('scrollTop',scrollTop, '判断条件', listElementMessage[i]+gTopH+interval)
          i === 0 ? lastTabSelect = i : lastTabSelect = i - 1
          tabSelect = lastTabSelect
          this.setState({
            tabSelect
          })
          break          
        }
      }
    }

    // 保存滚动
    this.setState({
      scrollTop,
    })
  }

  // 给需要滚动的子元素加key
  setKeyChildren = () => {
    const { galleryElements, prefixCls } = this.props
    // console.log('galleryElements---', galleryElements)
    if (galleryElements ) {
      return galleryElements.map((item, index) => {
        return (
          <div className={`${prefixCls}-child`} key={index}>
            {item}
          </div>
        )
      })
    } else {
      return (<div>暂无数据或者galleryElements所传递的不是react node 数组</div>)
    }

  }

  render () {
    const { barTab, prefixCls, barMinWidth, outBoxHeight } = this.props
    const tab = barTab && barTab.length > 0 && (
      barTab.map((item, index) => {
        const tabClassName = classNames({
          [`${prefixCls}-bar-list`]: true,
          [`${prefixCls}-bar-box-select`]: this.state.tabSelect === index,
        })
        return (
          <div className={tabClassName} key={index} onClick={e => this.handleTabClick(e, index)}>{item}</div>
        )
      })
    )
    
    return (
      <PerfectScrollbar style={{height: outBoxHeight}}>
        <div className={`${prefixCls}-out-box`}>
          <div className={`${prefixCls}-box`} ref={ galleryBox => this.galleryBox = galleryBox } onScroll={this.handleGalleryScroll}>
            <div className={`${prefixCls}`} ref={ gallery => this.gallery = gallery }>
              <div className={`${prefixCls}-top`} ref={gTop => this.gTop = gTop}>
                {this.props.galleryTop}
              </div>
              <div className={`${prefixCls}-bar`} ref={barTab => this.barTab = barTab }>
                <div className={`${prefixCls}-bar-box`} style={{ minWidth: barMinWidth }}>
                  {tab}
                </div>
              </div>
              <div className={`${prefixCls}-listBox`} ref={listBox => this.listBox = listBox}>
                {this.setKeyChildren()}
              </div>
            </div>
          </div>
          <div className={`${prefixCls}-out-box-barFixBox`} ref={barTabTwo => this.barTabTwo = barTabTwo }>
            <div className={`${prefixCls}-out-box-bar`}>
              <div className={`${prefixCls}-out-box-bar-box`} style={{ minWidth: barMinWidth }}>
                {tab}
              </div>
            </div>
          </div>
        </div>
      </PerfectScrollbar>
    )
  }
}

ScrollGalleryShow.propTypes = {
  barTab: PropTypes.array,
  tabSelect: PropTypes.number,
  galleryTop: PropTypes.element,
  galleryElements: PropTypes.arrayOf(PropTypes.element),
  prefixCls: PropTypes.string,
  barMinWidth: PropTypes.oneOfType([ PropTypes.string, PropTypes.number ]),
  outBoxHeight: PropTypes.number,
}
ScrollGalleryShow.defaultProps = {
  tabSelect: 0,
  prefixCls: 'cr-scrollgallery',
  barMinWidth: '90%',
  outBoxHeight: 600,
}
export default ScrollGalleryShow

examples/testOne.js
import React, { PureComponent } from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'
import ScrollGallery from 'cr-scrollgallery'
import '../assets/index.less';
import styles from '../assets/index.less'

const reactContainer = document.getElementById('__react-content');
const bodyContainer = document.getElementsByTagName('body')
bodyContainer[0].style.padding = '10px'

class App extends PureComponent {

  componentDidMount() {
    const enterDom = document.getElementById('enter')
    setTimeout(() => {
      const helloDom = document.createElement('div')
      helloDom.style.height = 300 + 'px'
      helloDom.style.border = '1px solid red'
      enterDom.appendChild(helloDom)
    }, 1000)
  }
  handleClick = () => {
    console.log(this.gallery)
    this.gallery.setInitScrollNum()
    console.log('click 0000')

  }
  
  render () {
    const galleryTop = (
      <div style={{border: '1px solid red', height: 300}}></div>
    )
    const arrayElement = () => {
      const one = (
        <div key='one'>
          <div style={{ height: 80}}>tab一</div>
          <div>dfdfjajdf</div>
          <div>dfdfjajdf</div>
          <div>dfdfjajdf</div>
          <div>dfdfjajdf</div>
          <div>dfdfjajdf</div>
          <div style={{border: '1px solid green', height: 450}}></div>
        </div>
      )
      const two = (
        <div key='two' id='enter'>
          <div style={{ height: 60}}>tab二</div>
          <div style={{border: '1px solid green', height: 450}}></div>
        </div>
      )
      const three = (
        <div key='three'>
          <div style={{ height: 70}}>tab三</div>
          <div>dfdfjajdf</div>
          <div>dfdfjajdf</div>
          <div>dfdfjajdf</div>
          <div style={{border: '1px solid green', height: 3450}}></div>
        </div>
      )

      return [one, two, three]
    }

    const elementTwo = () => {
      const data = {}
      const one = (<div  style={{height: 600, border: '1px solid red'}}>
        <div className={styles.title}>工商信息</div>
        <table className={styles.table}>
          <tr>
            <th>法定代表人</th>
            <td colSpan="3">{(data && data.OperName) || ''}</td>
          </tr>
          <tr>
            <th>注册资本</th>
            <td>{(data && data.RegistCapi) || ''}</td>
            <th>实缴资本</th>
            <td>{(data && data.aaa) || ''}</td>
          </tr>
          <tr>
            <th>经营状态</th>
            <td>{(data && data.aaa) || ''}</td>
            <th>成立日期</th>
            <td>{(data && data.StartDate) || ''}</td>
          </tr>
          <tr>
            <th>统一社会信用代码</th>
            <td>{(data && data.CreditCode) || ''}</td>
            <th>纳税人识别号</th>
            <td>{(data && data.aaa) || ''}</td>
          </tr>
          <tr>
            <th>注册号</th>
            <td>{(data && data.No) || ''}</td>
            <th>组织机构代码</th>
            <td>{(data && data.OrgNo) || ''}</td>
          </tr>
          <tr>
            <th>公司类型</th>
            <td>{(data && data.EconKind) || ''}</td>
            <th>所属行业</th>
            <td>{(data && data.aaa) || ''}</td>
          </tr>
          <tr>
            <th>核准日期</th>
            <td>{(data && data.aaa) || ''}</td>
            <th>登记机关</th>
            <td>{(data && data.BelongOrg) || ''}</td>
          </tr>
          <tr>
            <th>所属地区</th>
            <td>{(data && data.Province) || ''}</td>
            <th>英文名</th>
            <td>{(data && data.aaa) || '-'}</td>
          </tr>
          <tr>
            <th>曾用名</th>
            <td>{(data && data.aaa) || '-'}</td>
            <th>参保人数 </th>
            <td>{(data && data.aaa) || ''}</td>
          </tr>
          <tr>
            <th>人员规模</th>
            <td>{(data && data.aaa) || ''}</td>
            <th>营业期限</th>
            <td>{(data && data.aaa) || ''}</td>
          </tr>
          <tr>
            <th>企业地址</th>
            <td colSpan="3">{(data && data.Address) || ''}</td>
          </tr>
          <tr>
            <th>经营范围</th>
            <td colSpan="3">{(data && data.Scope) || ''}</td>
          </tr>
        </table>
      </div>)

      const two = (<div  style={{height: 600, border: '1px solid red'}}>
        <div className={styles.title}>其他信息</div>
        <table className={styles.table}>
          <tr>
            <th>法定代表人</th>
            <td colSpan="3">{(data && data.OperName) || ''}</td>
          </tr>
          <tr>
            <th>注册资本</th>
            <td>{(data && data.RegistCapi) || ''}</td>
            <th>实缴资本</th>
            <td>{(data && data.aaa) || ''}</td>
          </tr>
          <tr>
            <th>经营状态</th>
            <td>{(data && data.aaa) || ''}</td>
            <th>成立日期</th>
            <td>{(data && data.StartDate) || ''}</td>
          </tr>
          <tr>
            <th>统一社会信用代码</th>
            <td>{(data && data.CreditCode) || ''}</td>
            <th>纳税人识别号</th>
            <td>{(data && data.aaa) || ''}</td>
          </tr>
          <tr>
            <th>注册号</th>
            <td>{(data && data.No) || ''}</td>
            <th>组织机构代码</th>
            <td>{(data && data.OrgNo) || ''}</td>
          </tr>
          <tr>
            <th>公司类型</th>
            <td>{(data && data.EconKind) || ''}</td>
            <th>所属行业</th>
            <td>{(data && data.aaa) || ''}</td>
          </tr>
          <tr>
            <th>核准日期</th>
            <td>{(data && data.aaa) || ''}</td>
            <th>登记机关</th>
            <td>{(data && data.BelongOrg) || ''}</td>
          </tr>
          <tr>
            <th>所属地区</th>
            <td>{(data && data.Province) || ''}</td>
            <th>英文名</th>
            <td>{(data && data.aaa) || '-'}</td>
          </tr>
          <tr>
            <th>曾用名</th>
            <td>{(data && data.aaa) || '-'}</td>
            <th>参保人数 </th>
            <td>{(data && data.aaa) || ''}</td>
          </tr>
          <tr>
            <th>人员规模</th>
            <td>{(data && data.aaa) || ''}</td>
            <th>营业期限</th>
            <td>{(data && data.aaa) || ''}</td>
          </tr>
          <tr>
            <th>企业地址</th>
            <td colSpan="3">{(data && data.Address) || ''}</td>
          </tr>
          <tr>
            <th>经营范围</th>
            <td colSpan="3">{(data && data.Scope) || ''}</td>
          </tr>
        </table>
      </div>)

    return [one, two]
    }

    return (
      <div>
        {/* <div style={{width: 100, height: 100, border: '1px solid red'}} onClick={this.handleClick}></div> */}
        <div style={{height: 400, border: '10px solid #21c8be'}}>
          <ScrollGallery
              ref={gallery => this.gallery = gallery }
              barTab={['tab一', 'tab二', 'tab三']}
              // tabSelect={1}
              galleryTop={galleryTop}
              galleryElements={arrayElement()}
              barMinWidth={500}
              outBoxHeight={400}
          />
        </div>
      </div>
    )
  }
}


function render(container) {
  ReactDOM.render(
    <div>
      <h3>RCSCROLLGALLERY TEST</h3>
      <div><App /></div>               
    </div>, container
  )
}

render(reactContainer)